File: PathoFroh/src/main/java/com/patho/main/service/WorkPhaseService.java
package com.patho.main.service;

import com.patho.main.common.PredefinedFavouriteList;
import com.patho.main.model.patient.DiagnosisRevision;
import com.patho.main.model.patient.DiagnosisRevision.NotificationStatus;
import com.patho.main.model.patient.Task;
import com.patho.main.repository.TaskRepository;
import com.patho.main.util.helper.HistoUtil;
import com.patho.main.util.task.TaskStatus;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;

@Service
@Transactional
public class WorkPhaseService extends AbstractService {

    @Autowired
    private TaskRepository taskRepository;

    @Autowired
    private FavouriteListService favouriteListService;

    @Autowired
    private SlideService slideService;

    @Autowired
    private DiagnosisService diagnosisService;

    /**
     * Updates the status of the stating phase, if the phase has ended, true will be
     * returned.
     *
     * @param task
     */
    @Transactional
    public boolean updateStainigPhase(Task task) {
        // removing from staining list and showing the dialog for ending
        // staining phase
        if (TaskStatus.checkIfStainingCompleted(task)) {
            logger.trace("Staining phase of task (" + task.getTaskID()
                    + ") completed removing from staing list, adding to diagnosisList");
            // do nothin is done manual via dialog
            return true;
        } else {
            // reentering the staining phase, adding task to staining or
            // restaining list
            return false;
        }
    }

    /**
     * Start the staining phase, adds the task to the staining or restaining phase,
     * set staining completion date to 0
     * <p>
     * Error-Handling via global Error-Handler
     *
     * @param task
     */
    @Transactional
    public Task startStainingPhase(Task task) {

        if (task.getStainingCompletionDate() != null) {
            task.setStainingCompletionDate(null);
            task = taskRepository.save(task, resourceBundle.get("log.phase.staining.enter"));
        }

        if (!TaskStatus.hasFavouriteLists(task, PredefinedFavouriteList.StainingList,
                PredefinedFavouriteList.ReStainingList)) {
            if (TaskStatus.checkIfReStainingFlag(task)) {
                task = favouriteListService.removeTaskFromList(task, PredefinedFavouriteList.StainingList);
                task = favouriteListService.addTaskToList(task, PredefinedFavouriteList.ReStainingList);
            } else
                task = favouriteListService.addTaskToList(task, PredefinedFavouriteList.StainingList);
        }

        return task;
    }

    /**
     * Ends the staining phase, removes the task from the staining lists, sets
     * staining completion date to current time
     * <p>
     * Error-Handling via global Error-Handler
     *
     * @param task
     */
    @Transactional
    public Task endStainingPhase(Task task, boolean removeFromList) {
        slideService.completedStaining(task, true);
        task.setStainingCompletionDate(Instant.now());

        task = taskRepository.save(task, resourceBundle.get("log.phase.staining.end"));

        if (removeFromList)
            task = favouriteListService.removeTaskFromList(task, PredefinedFavouriteList.StainingList,
                    PredefinedFavouriteList.ReStainingList);

        return task;
    }

    /**
     * Starts the diagnosis phase, sets the time of diagnosis completion to zero and
     * adds the task to the diagnosis list.
     *
     * @param task
     */
    @Transactional
    public Task startDiagnosisPhase(Task task) {
        task.setDiagnosisCompletionDate(null);

        task = taskRepository.save(task, resourceBundle.get("log.phase.diagnosis.enter"));

        if (!TaskStatus.hasFavouriteLists(task, PredefinedFavouriteList.DiagnosisList,
                PredefinedFavouriteList.ReDiagnosisList)) {
            task = favouriteListService.addTaskToList(task.getId(), PredefinedFavouriteList.DiagnosisList);
        }

        return task;
    }

    /**
     * Ends the diagnosis phase, removes from diagnosis list and sets the diagnosis
     * time of completion.
     */
    public Task endDiagnosisPhase(Task task, boolean removeFromList, NotificationStatus status) {

        boolean change = false;
        // setting diagnosis compleation date if not set jet
        for (int i = 0; i < task.getDiagnosisRevisions().size(); i++) {
            DiagnosisRevision rev = HistoUtil.getNElement(task.getDiagnosisRevisions(), i);
            if (rev.getCompletionDate() == null) {
                task = diagnosisService.approveDiangosis(task, rev, status);
                change = true;
            }
        }

        if (change || task.getDiagnosisCompletionDate() == null) {
            task.setDiagnosisCompletionDate(Instant.now());
            task = taskRepository.save(task, resourceBundle.get("log.phase.diagnosis.end"));
        }

        if (removeFromList)
            task = favouriteListService.removeTaskFromList(task, PredefinedFavouriteList.DiagnosisList,
                    PredefinedFavouriteList.ReDiagnosisList);
        return task;
    }
}

File: PathoFroh/src/main/kotlin/com/patho/main/service/WorkPhaseService.kt
package com.patho.main.service

import com.patho.main.common.PredefinedFavouriteList
import com.patho.main.model.patient.NotificationStatus
import com.patho.main.model.patient.Task
import com.patho.main.repository.TaskRepository
import com.patho.main.util.exceptions.DiagnosisRevisionsNotCompletedException
import com.patho.main.util.exceptions.NotificationNotCompletedException
import com.patho.main.util.exceptions.StainingsNotCompletedException
import com.patho.main.util.task.TaskStatus
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant

@Service()
open class WorkPhaseService @Autowired constructor(
        private val taskRepository: TaskRepository,
        private val favouriteListService: FavouriteListService,
        private val slideService: SlideService,
        private val diagnosisService: DiagnosisService) : AbstractService() {

    /**
     * Start the staining phase, adds the task to the staining or restaining phase,
     * set staining completion date to 0
     */
    @Transactional
    open fun startStainingPhase(task: Task): Task {
        var tmp = task

        if (tmp.stainingCompleted) {
            tmp.stainingCompletionDate = null
            tmp = taskRepository.save(tmp, resourceBundle.get("log.phase.staining.enter", tmp), tmp.patient)
        }

        if (!TaskStatus.hasFavouriteLists(tmp, PredefinedFavouriteList.StainingList,
                        PredefinedFavouriteList.ReStainingList)) {
            if (TaskStatus.checkIfReStainingFlag(tmp)) {
                tmp = favouriteListService.removeTaskFromList(tmp, PredefinedFavouriteList.StainingList)
                tmp = favouriteListService.addTaskToList(tmp, PredefinedFavouriteList.ReStainingList)
            } else
                tmp = favouriteListService.addTaskToList(tmp, PredefinedFavouriteList.StainingList)
        }

        return tmp
    }

    /**
     * Ends the staining phase, removes the task from the staining lists, sets
     * staining completion date to current time
     */
    @Transactional
    open fun endStainingPhase(task: Task, removeFromList: Boolean): Task {
        var tmp = task

        // throws error if not all staings are completed
        if (!TaskStatus.checkIfStainingCompleted(task))
            throw  StainingsNotCompletedException()

        tmp.stainingCompletionDate = Instant.now()
        tmp = taskRepository.save(tmp, resourceBundle.get("log.phase.staining.end", tmp), tmp.patient)

        if (removeFromList)
            tmp = favouriteListService.removeTaskFromList(tmp, PredefinedFavouriteList.StainingList,
                    PredefinedFavouriteList.ReStainingList)

        return tmp
    }

    /**
     * Starts the reportIntent phase, sets the time of reportIntent completion to zero and
     * adds the task to the reportIntent list.
     */
    @Transactional
    open fun startDiagnosisPhase(task: Task): Task {
        var tmp = task
        tmp.diagnosisCompletionDate = null

        tmp = taskRepository.save(tmp, resourceBundle.get("log.phase.diagnosis.enter", tmp), tmp.patient)

        if (!TaskStatus.hasFavouriteLists(tmp, PredefinedFavouriteList.DiagnosisList,
                        PredefinedFavouriteList.ReDiagnosisList)) {
            tmp = favouriteListService.addTaskToList(tmp.id, PredefinedFavouriteList.DiagnosisList)
        }

        return tmp
    }

    /**
     * Ends the diagnosis phase, removes from diagnosis list and sets the diagnosis
     * time of completion. If not all diagnoses had been marked as completed a DiagnosisRevisionsNotCompletedException is
     * thrown
     */
    @Transactional
    open fun endDiagnosisPhase(task: Task, removeFromList: Boolean): Task {
        var tmp = task

        if (tmp.diagnosisRevisions.any { !it.completed })
            throw  DiagnosisRevisionsNotCompletedException()

        tmp.diagnosisCompletionDate = Instant.now()
        tmp = taskRepository.save(tmp, resourceBundle.get("log.phase.diagnosis.end", tmp), tmp.patient)

        if (removeFromList)
            tmp = favouriteListService.removeTaskFromList(tmp, PredefinedFavouriteList.DiagnosisList,
                    PredefinedFavouriteList.ReDiagnosisList)
        return tmp
    }

    /**
     * Starting notification phase and sets the notificationCompletionDate to null. The task is also added to the
     * notification list.
     */
    @Transactional
    open fun startNotificationPhase(task: Task): Task {
        var tmp = task
        tmp.notificationCompletionDate = null
        tmp = taskRepository.save(tmp, resourceBundle.get("log.phase.notification.enter", tmp), tmp.patient)
        return favouriteListService.addTaskToList(tmp, PredefinedFavouriteList.NotificationList)
    }

    /**
     * Ends the notification phase. If not all notification are marked as completed, a NotificationNotCompletedException
     * is thrown.
     */
    @Transactional
    open fun endNotificationPhase(task: Task, removeFromList: Boolean): Task {
        var tmp = task

        if (tmp.diagnosisRevisions.any { !it.isNotified })
            throw NotificationNotCompletedException()

        tmp.notificationCompletionDate = Instant.now()
        tmp = taskRepository.save(tmp, resourceBundle.get("log.phase.notification.end", tmp), tmp.patient)

        if (removeFromList)
            tmp = favouriteListService.removeTaskFromList(tmp, PredefinedFavouriteList.NotificationList,
                    PredefinedFavouriteList.ReDiagnosisList)

        return tmp
    }
}